nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-ssl-autoloading-utils/src/main/java/org/apache/nifi/autoload/PrivateKeyInfo.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.autoload;
+
+import java.util.Objects;
+
+public class PrivateKeyInfo implements Comparable<PrivateKeyInfo> {
+    private static final String COMPARISON_FORMAT = "%s-%s-%s-%s";
+
+    /**
+     * Creates a PrivateKeyInfo containing the state we want to track.
+     * @param subjectDN The subject DN of the cert
+     * @param issuerDN The issuer DN
+     * @param alias The alias in the keystore
+     * @param issuerSerialNumber Optional serial number of the issuer, if the issuer's certificate was found
+     *                           in the cert chain
+     */
+    public PrivateKeyInfo(String subjectDN, String issuerDN, String alias, String issuerSerialNumber) {
+        this.subjectDN = subjectDN;
+        this.issuerDN = issuerDN;
+        this.alias = alias;
+        this.issuerSerialNumber = issuerSerialNumber;
+    }
+
+    private String subjectDN;
+    private String issuerDN;
```

Comment:
Is there a reason for using `String` as opposed to `X500Principal` for these properties?  Using the `X500Principal` class follows the property types from `X509Certificate` and preserves the original binary form as opposed to the string representation.  Also recommend renaming them to `subjectPrincipal` and `issuerPrincipal` if the type is changed.

Reply:
Great point, and this would remove the need for the issuerSerialNumber as well.